### PR TITLE
[ICDS Dashboard] Added vhsnd monthly report

### PR DESCRIPTION
**Pull request by challabeehyv**
https://app.asana.com/0/0/1158279992597553/f

### CODE REVIEW

**Comment by mkangia** [comment]
@calellowitz  can you confirm on the commits? There are two commits that are not related to code changes, and i am not sure if they are good to merge.

The code changes in the third commit were already reviewed in the [closed PR](https://github.com/dimagi/commcare-hq/pull/26789)

**Review by mkangia — APPROVED** [review]
Code changes are good to go, not sure about the commit trail, so asked for that for confirmation.

File: custom/icds_reports/data_pull/exporter.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def __init__(self, slug_or_file, db_alias, month, location_id):
     @prevent_parallel_execution(DATA_PULL_CACHE_KEY)
     def export(self):
         zip_file_name = "%s-DataPull.zip" % self.data_pull_obj.name
-        with zipfile.ZipFile(zip_file_name, mode='a') as z:
+        with zipfile.ZipFile(zip_file_name, mode='w') as z:
```
Comment (Rohit25negi) [inline custom/icds_reports/data_pull/exporter.py]: Why is this changed?

File: custom/icds_reports/data_pull/data_pulls.py [hunk 2]
```diff
@@ -193,3 +203,86 @@ def _dump_consolidated_data(result):
...
+        headers.extend(dates)
+        headers.append('Grand Total')
+        return headers, dates
+
+    def _setup_state_results_filestream(self, headers, result):
+        state_results = defaultdict(dict)
+        for state_name in result.keys():
+            # constructing and mapping writers to state names
+            if state_name in state_results:
+                raise DuplicateStateResult("%s is twice in results" % state_name)
```
Comment (Rohit25negi) [inline custom/icds_reports/data_pull/data_pulls.py]: result is a dict? how would it have same state name twice? i.e multiple keys with same name twice?

**Review by Rohit25negi — CHANGES_REQUESTED** [review]
Looks good. Left two questions.
Is this tested?

File: custom/icds_reports/data_pull/exporter.py (hunk 1 above)
Comment (mkangia) [reply custom/icds_reports/data_pull/exporter.py]: since we are creating a new file "w" would the write thing to avoid appending to an existing file?

File: custom/icds_reports/data_pull/data_pulls.py (hunk 2 above)
Comment (mkangia) [reply custom/icds_reports/data_pull/data_pulls.py]: That was my suggestion and what you are saying sounds correct to me. If that check is needed it should be moved to a better place.

Comment (challabeehyv) [reply custom/icds_reports/data_pull/data_pulls.py]: Yes. the check seems unnecessary. It will from the earlier approach where the workbook construction happening on the data population loop. removed it

**Comment by challabeehyv** [comment]
> Looks good. Left two questions.
> Is this tested?

[VHSND monthly report-DataPull.zip](https://github.com/dimagi/commcare-hq/files/4322747/VHSND.monthly.report-DataPull.zip)

File: custom/icds_reports/data_pull/exporter.py (hunk 1 above)
Comment (Rohit25negi) [reply custom/icds_reports/data_pull/exporter.py]: how was being handled previously?

Comment (mkangia) [reply custom/icds_reports/data_pull/exporter.py]: It didn't really cause any issue earlier because any request was not repeated but good to keep it new each time.
Also zip file name i think has the timestamp as well so reusing a file won't happen but have a new file sounds like the right thing to do.

Comment (Rohit25negi) [reply custom/icds_reports/data_pull/exporter.py]: or you are saying previously, we were using it once so it was not appending and if we were running it twice we used to delete the old one?

Comment (mkangia) [reply custom/icds_reports/data_pull/exporter.py]: Actually we have never run this more than once. Plus we remove the file after use
https://github.com/dimagi/commcare-hq/blob/3a8e462914827c351efe10a1d36e8cbfc67cf3e5/custom/icds/tasks/data_pulls.py#L57

**Comment by mkangia** [comment]
@Rohit25negi @challabeehyv 
I think this is good to merge now unless @Rohit25negi has some pending feedback

**Review by Rohit25negi — APPROVED** [review]

**Comment by Rohit25negi** [comment]
Looks good to me. @mkangia  commit trail looks just merge with master and the final diff is only the essentials so I think this is good to go.

**Comment by mkangia** [comment]
I was checking with others about the extra commits.
Are they because you need to merge dimagi:master to your fork @challabeehyv ?

**Comment by challabeehyv** [comment]
> I was checking with others about the extra commits.
> Are they because you need to merge dimagi:master to your fork @challabeehyv ?

yes. That is correct

**Comment by mkangia** [comment]
Would appreciate if that is followed from now on, for cleaner PRs
@challabeehyv 
fyi @Rohit25negi
